lib/Auth/utils.php

```diff
@@ -1,124 +1,134 @@
-		);
-
-		$body = self::curl($params, null, $serverconfig, "/auth", "POST");
-		try {
-			$result = $body->result;
-			$value = $result->value;
-			$token = $value->token;
-		} catch (Exception $e) {
-			throw new SimpleSAML_Error_BadRequest("privacyIDEA: We were not able to read the response from the PI server");
-		}
-		return $token;
-	}
-
-	/**
-	 * This function can edit the state to enter the needed token types for a user.
-	 * The booleans 'use_u2f' and 'use_otp' will be added.
-	 *
-	 * @param array $state
-	 * The state is needed to be changed in this function
-	 *
-	 * @param JSON $body
-	 * The body contains the multi_challenge which will be used to check which token types are used.
-	 *
-	 * @return mixed
-	 * The modified state will be returned. It now contains the token types for the user.
-	 */
-	public function checkTokenType($state, $body) {
-		$detail = $body->detail;
-		$multi_challenge = $detail->multi_challenge;
-		$use_u2f = false;
-		$use_otp = false;
-		$use_push = false;
-		for ($i = 0; $i < count($multi_challenge); $i++) {
-		    switch ($multi_challenge[$i]->type) {
+class sspmod_privacyidea_Auth_utils
+{
+
+    /**
+     * Perform a request against privacyIDEA.
+     *
+     * @param array $params All params, which are needed for the http request (e.g. user, pass, realm, etc.)
+     * @param array $headers The headers for the http request (e.g. authentication token)
+     * @param array $serverconfig The whole configuation for the server (e.g. url, verify host, verify peer)
+     * @param string $api_endpoint This is the path for the request (e.g. /validate/samlcheck)
+     * @param string $http_method Some requests need POST or GET method. This can be entered here.
+     * @return object We will return the JSON decoded body, because all the requests need different data.
+     * @throws SimpleSAML_Error_BadRequest
+     */
+    public static function curl($params, $headers, $serverconfig, $api_endpoint, $http_method)
+    {
+        assert('array' === gettype($params));
+        assert('array' === gettype($headers));
+        assert('array' === gettype($serverconfig));
+        assert('string' === gettype($api_endpoint));
+        assert('string' === gettype($http_method));
+
+        $curl_instance = curl_init();
+        $url = $serverconfig['privacyideaserver'] . $api_endpoint;
+
+        curl_setopt($curl_instance, CURLOPT_URL, $url);
+        curl_setopt($curl_instance, CURLOPT_HEADER, true);
+        if ($headers) {curl_setopt($curl_instance, CURLOPT_HTTPHEADER, $headers);}
+        curl_setopt($curl_instance, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl_instance, CURLOPT_USERAGENT, USERAGENT);
+        if ($http_method === "POST") {
+            curl_setopt($curl_instance, CURLOPT_POST, true);
+            curl_setopt($curl_instance, CURLOPT_POSTFIELDS, $params);
+        } elseif ($http_method === "GET") {
+            $params_str = '?';
+            foreach ($params as $key => $value) {
```

Comment:
Are the parameters url encoded?

Reply:
Nope. This still needs changing, but this PR is already too complex, I had to stop somewhere (especially since privacyIDEA handles url decoding inconsistently, so there is no easy fix for this).